File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/atomic/SimpleShape2D.java
package com.greentree.commons.geometry.geom2d.atomic;

import com.greentree.commons.math.MathLine1D;
import com.greentree.commons.math.vector.AbstractVector2f;
import com.greentree.commons.math.vector.Vector2f;
import com.greentree.commons.math.vector.Vector3f;
import com.greentree.commons.util.iterator.IteratorUtil;
import org.joml.Matrix2f;
import org.joml.Matrix3f;

public interface SimpleShape2D {

    default MathLine1D getProjection(Matrix3f model, AbstractVector2f normal) {
        var v = normal.times(new Matrix2f(model));
        var len = v.length();
        v = v.normalize(1);
        var ml = getProjection(v);
        return new MathLine1D(ml.min * len, ml.max * len);
    }

    MathLine1D getProjection(AbstractVector2f normal);

    default boolean isInsade(Matrix3f model, AbstractVector2f point) {
        var invModelMatrix = new Matrix3f();
        model.invert(invModelMatrix);
        final var v = new Vector3f(point, 1).times(invModelMatrix);
        return isInsade(v.xy());
    }

    boolean isInsade(AbstractVector2f point);

    default AbstractVector2f min(Matrix3f model, AbstractVector2f point) {
        var invModelMatrix = new Matrix3f();
        model.invert(invModelMatrix);
        final var v = new Vector3f(point, 1).times(invModelMatrix);
        return new Vector3f(min(v.xy()), 1).times(model).xy();
    }

    AbstractVector2f min(AbstractVector2f point);

    default Iterable<? extends AbstractVector2f> normals(Matrix3f model) {
        return IteratorUtil.map(normals(), n -> new Vector2f(n).times(new Matrix2f(model)));
    }

    Iterable<? extends AbstractVector2f> normals();

    default Iterable<? extends AbstractVector2f> points(Matrix3f model) {
        return IteratorUtil.map(points(), n -> new Vector3f(n, 1).times(model).xy());
    }

    Iterable<? extends AbstractVector2f> points();

}

File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/atomic/SimpleShape2D.kt
package com.greentree.commons.geometry.geom2d.atomic

import com.greentree.commons.math.MathLine1D
import com.greentree.commons.math.vector.AbstractVector2f
import com.greentree.commons.math.vector.Vector3f
import org.joml.Matrix2f
import org.joml.Matrix3f
import org.joml.Matrix3fc

interface SimpleShape2D {

	fun projection(normal: AbstractVector2f): MathLine1D

	fun isInside(point: AbstractVector2f): Boolean

	fun nearestPoint(point: AbstractVector2f): AbstractVector2f

	fun normals(): Iterable<AbstractVector2f>

	fun points(): Iterable<AbstractVector2f>
}

fun SimpleShape2D.projection(model: Matrix3fc, normal: AbstractVector2f): MathLine1D {
	var v = normal * Matrix2f(model)
	val len = v.length()
	v = v.normalize(1)
	val ml = projection(v)
	return MathLine1D(ml.min * len, ml.max * len)
}

fun SimpleShape2D.isInside(model: Matrix3fc, point: AbstractVector2f): Boolean {
	val invModelMatrix = Matrix3f()
	model.invert(invModelMatrix)
	val v = Vector3f(point, 1f).times(invModelMatrix)
	return isInside(v.xy())
}

fun SimpleShape2D.nearestPoint(model: Matrix3fc, point: AbstractVector2f): AbstractVector2f {
	val invModelMatrix = Matrix3f()
	model.invert(invModelMatrix)
	val v = Vector3f(point, 1f).times(invModelMatrix)
	return Vector3f(nearestPoint(v.xy()), 1f).times(model).xy()
}

fun SimpleShape2D.normals(model: Matrix3fc): Iterable<out AbstractVector2f> {
	return normals().map { it * Matrix2f(model) }
}

fun SimpleShape2D.points(model: Matrix3fc): Iterable<AbstractVector2f?> {
	return points().map { it * model }
}